pac4j-saml-config-db/src/main/java/org/pac4j/saml/dbclient/DbLoadedSamlClients.java

```diff
@@ -0,0 +1,197 @@
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.dbclient.dao.api.SamlClientDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An extension to base {@link Clients}, which works also with clients whose definition is stored in a database. 
+ *
+ * @author jkacer
+ * @since 1.9.0
+ */
+public class DbLoadedSamlClients extends Clients {
+
+    /** SLF4J logger. */
+    private final Logger logger = LoggerFactory.getLogger(DbLoadedSamlClients.class);
+    
+    /** DAO reading definitions of SAML clients from a database. */
+    private SamlClientDao samlClientDao;
+
+    /** Clients set explicitly using a setter. */
+    private List<Client> explicitClients;
+    
+    // ------------------------------------------------------------------------------------------------------------------------------------
+
+    
+	/**
+	 * Creates a new client group with no explicit clients.
+	 */
+	public DbLoadedSamlClients() {
+		super();
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(Client... clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with a single explicit client.
+	 * 
+	 * @param client
+	 *            The client.
+	 */
+	public DbLoadedSamlClients(Client client) {
+		super(client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(List<Client> clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client... clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param client
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client client) {
+		super(callbackUrl, client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 *
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, List<Client> clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Initializes all clients by loading them from the database.
+	 * 
+	 * @see org.pac4j.core.util.InitializableWebObject#internalInit(org.pac4j.core.context.WebContext)
+	 */
+	@Override
+	protected void internalInit(final WebContext context) {
+        // Remove all clients first if re-initialized

```

Comment:
@leleuj Sorry, but I have another question / comment.
The part about the configuration is clear, I have already done that.

However, a problem arises when I move to Clients and SAML2Client. What you wrote above would, in my opinion, work well if at the beginning we had the Clients object initialized with a list of SAML2Client-s and their not-yet-loaded DatabaseSAML2Configurations. Then, it would work as you suggest:
- Clients.findClient(name) - finds it OK by a known static name
- Clients.internalInit() - nothing important here
- SAML2Client.internalInit() - We try to initialize the configuration here, using a known name, and call DatabaseSAML2Configuration.init(clientName, context)
- Loaded OK --> everything OK

But it does not work if you have an unknown set of client definitions in the database and you don't know in advance:
- how many of them you have,
- what their names are.
You cannot create a new SAML2Client and its configuration in existing Clients because this class knows nothing about SAML clients etc. This means you will never find the client because Clients.findClient() iterates over a known list of clients. But you don't have the client in the list - Clients do not have this information because they don't know anything about SAML / SAML in DB.

That is why I first created the database-specific version of Clients. I agree with you that this is not nice and it should go away. That's why I proposed the dynamic "additional client contributor", which would tell Clients that there are additional clients that should be taken into account (and loaded later as you suggest).

Or was your intention to have the DB clients defined statically (as I described above, e.g. in Spring or simply Java) and just have all properties loaded dynamically?

Or maybe I haven't fully understood it yet?